packages/localize-tools/src/modes/transform.ts

```diff
@@ -345,10 +353,10 @@ class Transformer {
     if (ts.isTemplateExpression(template)) {
       for (const span of template.templateSpans) {
         // TODO(aomarks) Support less brittle/more readable placeholder keys.
-        const key = this.sourceFile.text.slice(
-          span.expression.pos,
-          span.expression.end
-        );
+        const key = this.sourceFile.text
+          .slice(span.expression.pos, span.expression.end)
+          .replace(/{ /g, '{')
```

Comment:
Curious about what's going on here, why is it needed? Seems like there could be some cases where this could break, such as `${"{ "}` (a literal string expression that would match the same regex).

Reply:
As we extract from the source code we run into this part https://github.com/lit/lit/blob/main/packages/localize-tools/src/program-analysis.ts#L332. If we get an object literal within an expression (like an options object to an inner `msg` call), the typescript printer will add spaces after/before opening/closing brackets regardless of what the source had and we store that in the `untranslatable` portion of a placeholder.

As we transform, we reconstruct the message by stitching together the translation contents, including the placeholder (which now includes bracket spaces). If the source code does not have bracket spaces within expressions, the transform errors as we can't find a matching expression now.

I'll try to think on it more but a way to make them congruent without forcing these replaces isn't coming to me right now.

Reply:
If we know the expression in the translation will always have bracket spaces (and I don't see an option for that in the typescript printer), we could take the source expression and run it through prettier with `bracketSpacing: true` to make them consistent.

Reply:
I see! Nicely sleuthed.

I'm thinking we might want to switch from using `printer.printNode` (https://github.com/lit/lit/blob/13b0a411a3c002540a687d0d70b499f801bf39e6/packages/localize-tools/src/program-analysis.ts#L332) to using the same source text extraction approach that we use in `transform.ts` (https://github.com/lit/lit/blob/13b0a411a3c002540a687d0d70b499f801bf39e6/packages/localize-tools/src/modes/transform.ts#L348). So:

```ts
file.text.slice(span.expression.pos, span.expression.end);
```

Seems like that would give us identically formatted text in both places.

Reply:
Hah, yeah that does it. The only concern for me then is that this will change the `extract` behavior as well and potentially change the placeholders in the extracted xliff/xlb. If a user originally had no bracket spaces, they'd have to re-extract and update the placeholders to match the new source for any already translated portions.

Reply:
That's true. I suppose the other option is to use `printer.printNode` in that transform mode spot then. Should have the same effect.